asim: measure thrashing

tbg opened the pull request:
This commit adds a thrashing benchmark to each stat output. "Thrashing" can be
measured in many ways, but here we measure, for each per-store timeseries, a
trend-discounting total variation. Roughly speaking, this computes the upward and
downward total variation separately, then interpolates between the minimum of both
and their sum. When one direction clearly dominates the other, it biases toward the
minimum. When up and down are roughly comparable, it returns something approximating
the sum - which is the "vanilla" total variation. In other words, thrashing is low when the
time series has a clear trend, and it is high when the series moves around but doesn't
seem to be trending anywhere.

When displaying the resulting value, we additionally normalize (divide by) the "range" of values
seen (i.e. the max minus the min), and print the result as a percentage.

For
example, the time series [0, 1, 0, 1, 0] has an upward and downward variation of 2. Neither dominates, so the trend-adjusted total variation equals the total variation 4. The values are all in `[0,1]`, so the range is 1. The resulting trend-discounting total variation percentage would be `tdtv/range = 4/1 = 400%`.

Intuitively, this means that the time series "sweeps" across the range
four times. Additional oscillations would drive this number up further.

In contast, the series `[0, 1, 0, 10, 100, 1000]` is mostly increasing, with the exception of the small downward jump between the second and third datapoints. Consequently, the trend-discounting total variation would (roughly) equal only the downward total variation, which equals 1. The range is 1000, so the percentage in this case would be `1/1000 = 0.1%`.

When printing the thrashing across a set of stores (but for a fixed stat),
instead of computing the range within each store's time series, we compute it
across the entire set. This makes it more likely that we discover the "true
range".

All of this can be made arbitrarily more complicated (and additional
normalization by the number of runs or ticks may be appropriate for some use
cases), but this should be a good start.

Epic: CRDB-49117
Release note: none

tbg committed:
asim: measure thrashing

This commit adds a thrashing benchmark to each stat output. "Thrashing" can be
measured in many ways, but here we measure, for each per-store timeseries, the
total variation (but excluding the first monotonic run). When displaying the
resulting value, we additionally normalize (divide by) the "range" of values
seen (i.e. the max minus the min), and print the result as a percentage. For
example, the time series [0, 1, 0, 1, 0, 1] would be assigned a thrashing
measurement of 200%, because:

- the range is 1 (two values are at most 1 apart from each other)
- there are three runs, each [0, 1], but we're ignoring the first one
- the remaining two runs contribute total variation of 1, each
- (1+1)/1 = 2 = 200%.

Intuitively, this means that the time series "sweeps" across the range
twice. Additional oscillations would drive this number up further.

When printing the thrashing across a set of stores (but for a fixed stat),
instead of computing the range within each store's time series, we compute it
across the entire set. This makes it more likely that we discover the "true
set".

All of this can be made arbitrarily more complicated (and additional
normalization by the number of runs or ticks may be appropriate for some use
cases), but this should be a good start.

Epic: CRDB-49117
Release note: none

tbg committed:
history: add more thrashing test cases

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152506)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_fulldisk.txt` line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OYa6V4229FhyGKSzptW:-OYa6V4229FhyGKSzptX:bs4cati) ([raw file](https://github.com/cockroachdb/cockroach/blob/c5155beafb6d4cca9a6a8c6af0b9dd50e2fb2399/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_fulldisk.txt#L30)):*
> ```help
> replicas#1: first: [s1=300, s2=300, s3=300, s4=300, s5=300] (stddev=0.00, mean=300.00, sum=1500)
> replicas#1: last:  [s1=322, s2=328, s3=327, s4=323, s5=200] (stddev=50.05, mean=300.00, sum=1500)
> replicas#1: thrash_pct: [s1=265%, s2=227%, s3=261%, s4=258%, s5=140%]  (sum=1152%)
> ```

![image.png](https://files.reviewableusercontent.io/images/reviewable/5076964/EXjm4-upLHW_B5FfPUz4emvx/image.png)

![image copy 1.png](https://files.reviewableusercontent.io/images/reviewable/5076964/-s9J4iUA8P38Wwhwlfg9T8wZ/image.png)

A few examples of what the corresponding graphs look like.

It feels like there should be a way to identify the "general direction" of where the equilibrium is (if one is reached) and to penalize only total variation that moves away from the equilibrium.

See monotonic_with_hiccup in the next commit. I'll likely play with this more before merging.


<!-- Sent from Reviewable.io -->

tbg committed:
asim: improve thrashing measurement

The previous mechanism was "total variation in all but the first run", but this
didn't do well on examples in which the series was basically monotonic but with
an early outlier (in which case the bulk of the "good monotonic function" would
also be tallied up as thrashing).

This commit introduces and adopts a "trend-discounting total variation".
Basically, it looks at the relative size of the upward and downward variation
and, if one dominates, discounts it in the computed value.

This should provide much better results. For example, in one instance from
example_liveness, we see the computed thrashing measure drop from 145% to 6%,
reflecting that most replica movements "went in the correct direction", which
wasn't previously captured properly.

tbg commented:
I added a commit that addresses the issue in my last comment, and updated the PR description with the latest approach.

tbg committed:
asim: improve thrashing stringer

tbg committed:
asim: change tdtv to support scaling property

When reviewing testdata changes, I realized that sometimes the trend-adjusted
total variation could increase even though the up and down variation had
independently become smaller. This behavior is uninutitive and undesirable, and
I traced it back to the "boosting" of the interpolation parameter I had
introduced in an earlier commit. This situation is now rectified.

tbg committed:
asim: skip prefix of zeroes in thrashing computation

This works around unexpected variation being detected when various timeseries
first get populated.

Reviewing the datadriven diffs shows that most thrashing percentages go down.
A few have increased, since they previously included these zero values into
the computation of the range of the time series. Now that these are stripped,
the denominator in the conversion from trend-discounting total variation becomes
smaller, resulting in increasing tdtv percentages. These changes are positive,
since the variation was always there, but previously masked.

tbg committed:
asim: add TestExploreTDTV

This can be used to understand how the tdtv approaches
its final value as it is repeatedly evaluated over longer
prefixes of the underlying metric.

tbg committed:
history: regenerate TestThrashing

tbg committed:
history: minor tweaks to TestThrashing

sumeerbhola reviewed (dismissed):
> @sumeerbhola reviewed 5 of 37 files at r1, 1 of 2 files at r2, 31 of 36 files at r4, 6 of 22 files at r6, 12 of 16 files at r7, 2 of 3 files at r8, 8 of 11 files at r9, 3 of 3 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152506)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @wenyihu6)

___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 104 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBVYaE5NQzEj8JFzaW:-OZBVYaE5NQzEj8JFzaX:b-7am27n) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/thrashing.go#L104)):*
>  For the endpoint 1, tdtv=2min(tu,td)

I assumed this was referring to the exponent, so I tried to simplify the resulting equation:
(min/max)*(tu+td) + (max-min)*min/max = (min*tu + min*td + max*tu + max*td + max*min - min^2)/max

say min = tu and max=td
(tu^2 + tu*td + tu*td + td^2 + tu*td - tu^2)/td  = 3*tu + td

which isn't equal to 2min = 2*tu

Did I make a mistake?
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_fulldisk.txt` line 27 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OYkmagiE2CvclpyQDoQ:-OYkmagjByYYtkvXj9VS:bl4zo4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6c5ef671acfb1e3c60c99c1727290ae64476c46/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_fulldisk.txt#L27)):*
> ```help
> disk_fraction_used#1: first: [s1=0.20, s2=0.20, s3=0.20, s4=0.20, s5=1.05] (stddev=0.34, mean=0.37, sum=2)
> disk_fraction_used#1: last:  [s1=0.30, s2=0.30, s3=0.30, s4=0.30, s5=0.95] (stddev=0.26, mean=0.43, sum=2)
> disk_fraction_used#1: thrash_pct: [s1=25%, s2=20%, s3=24%, s4=24%, s5=98%]  (sum=190%)
> ```

why is the thrashing % of {s1, s2, s3, s4} not close to 0 given they look monotonically increasing?
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 31 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBQrFo2LT1XhDeUvw0:-OZBQrFo2LT1XhDeUvw1:bmhnk2l) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/thrashing.go#L31)):*
> ```Go
> 	// normTV is a normalization factor for the total variation. By default, it is
> 	// initialized to the range of the input values, i.e. max - min (or 1 if max
> 	// == min). `tv/normTV` then measures how many times thrashing has "swept out"
> ```

should this be `tdtv/normTV`?
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 89 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBOgWy92wKSQFPFm2Z:-OZBOgWy92wKSQFPFm2_:b-978uo5) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/thrashing.go#L89)):*
> ```Go
> // Properties:
> // - symmetric in tu, td
> // - min(u,d) <= tdtv(tu,td) <= tu+td
> ```

what are `u` and `d` in `min(u,d)`?
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 91 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBOoUrCA5NpUbiEirV:-OZBOoUrCA5NpUbiEirW:bgsnqtp) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/thrashing.go#L91)):*
> ```Go
> // - min(u,d) <= tdtv(tu,td) <= tu+td
> // - tdtv(u,0) = tdtv(0,u) = 0
> // - tdtv(u,u) = 2t
> ```

what is t in 2t?
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 105 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBPgJ17JpothM05uSe:-OZBPgJ17JpothM05uSf:b-wugmjm) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/thrashing.go#L105)):*
> ```Go
> 	// 0.8 is somewhat arbitrary, but gives a reasonable trade-off.
> 	alpha := math.Pow(frac, 0.8)
> 	return alpha*(tu+td) + (1-alpha)*tmin
> ```

I'm trying to figure out the intuition here. As tu and td diverge, the frac will become closer to 0, so alpha will also become smaller. This will cause the weight to be mostly on tmin, which will be small. This sounds good, but I am curious what lead to this series of equations.
___
*[`pkg/kv/kvserver/asim/history/history.go` line 70 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBNm7Z7oft3Fbc6S0O:-OZBNm7Z7oft3Fbc6S0P:b52a9um) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/history.go#L70)):*
> ```Go
> 	ths := make(ThrashingSlice, numStores)
> 	for storeIdx := range vsByStore {
> 		// HACK: we remove leading zeroes before computing the thrashin This works
> ```

nit: thrashing.


<!-- Sent from Reviewable.io -->

tbg committed:
asim: fix up a comment

tbg committed:
history: update a comment

I think I first wrote it before improving the thrashing measure to use the
trend-discounting total variation. There was also a sentence in there that I
did not understand any more, so I removed it.

tbg committed:
history: improve variable names and comments

tbg committed:
history: fix TestExploreTDTV

It was busy looping and not doing at all what it was supposed to.

tbg committed:
history: fix filename

tbg committed:
history: rework the tdtv

Better explanations and rationales for the various invariants. Improved the
computation and highlighted its origins and alternatives.

tbg committed:
asim: add mma-and-count to example_fulldisk

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152506)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @sumeerbhola and @wenyihu6)

___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 31 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBQrFo2LT1XhDeUvw0:-O_MmpcvCRVu7lkEw-lS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/thrashing.go#L31)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

should this be `tdtv/normTV`?
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 89 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBOgWy92wKSQFPFm2Z:-O_Mo1-m7hOvamFBxD23:bmth19j) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/thrashing.go#L89)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

what are `u` and `d` in `min(u,d)`?
</blockquote></details>

Cleaned this up, should've been min(tu,td).
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 91 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBOoUrCA5NpUbiEirV:-O_Mo2lo1jgQpAm0AcLc:b-xtlksa) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/thrashing.go#L91)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

what is t in 2t?
</blockquote></details>

Cleaned this up. Should've been 2u.
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 104 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBVYaE5NQzEj8JFzaW:-O_MkdBM02NwFm9iQDav:b-tjcg0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/thrashing.go#L104)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

>  For the endpoint 1, tdtv=2min(tu,td)

I assumed this was referring to the exponent, so I tried to simplify the resulting equation:
(min/max)*(tu+td) + (max-min)*min/max = (min*tu + min*td + max*tu + max*td + max*min - min^2)/max

say min = tu and max=td
(tu^2 + tu*td + tu*td + td^2 + tu*td - tu^2)/td  = 3*tu + td

which isn't equal to 2min = 2*tu

Did I make a mistake?
</blockquote></details>

Yep, it's for p=1.

![image.png](https://files.reviewableusercontent.io/images/reviewable/5076964/nkHfrYVkhPIjPvIV-MUlJick/image.png)

```
(t_{min}+t_{max}) * \alpha + t_{min}*(1-\alpha)) &= (t_{min}+t_{max}) * \frac{t_{min}}{t_{max}} + t_{min}*(1-\frac{t_{min}}{t_{max}})) \\
& = \frac{t_{min}^2}{t_{max}} + t_{min} + t_{min} - \frac{t_{min}^2}{t_{max}} \\
& = 2*t_{min}
```
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 105 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OZBPgJ17JpothM05uSe:-O_QsInYAEJSZXCNPLHy:b-j3gv45) ([raw file](https://github.com/cockroachdb/cockroach/blob/40f16117d18093bc2a467d259be5e270781c5765/pkg/kv/kvserver/asim/history/thrashing.go#L105)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I'm trying to figure out the intuition here. As tu and td diverge, the frac will become closer to 0, so alpha will also become smaller. This will cause the weight to be mostly on tmin, which will be small. This sounds good, but I am curious what lead to this series of equations.
</blockquote></details>

I reworked the commentary and presentation here. Curious if it is more approachable now.
___
*[`pkg/kv/kvserver/asim/tests/testdata/non_rand/example_fulldisk.txt` line 27 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-OYkmagiE2CvclpyQDoQ:-O_QuEt2BvGdSgdF3hUS:b3b6u4s) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6c5ef671acfb1e3c60c99c1727290ae64476c46/pkg/kv/kvserver/asim/tests/testdata/non_rand/example_fulldisk.txt#L27)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

why is the thrashing % of {s1, s2, s3, s4} not close to 0 given they look monotonically increasing?
</blockquote></details>

![image.png](https://files.reviewableusercontent.io/images/reviewable/5076964/ZqF1ZOhD-5ZGTPytbI7jSTx0/image.png)

They actually thrash a lot, when normalized against themselves they are in the ~230% range. There is approx. 0.2 of upward variation vs. 0.1 of downward variation, so a ratio between them of 0.5, so the tdtv is still roughly comparable to the total variation of 0.3. Because the shared value range between the timeseries is roughly 1 (we see 0% and 100% disk utilization across the stores), 0.3 is also roughly the final score - around 30% thrashing.

They key is that they're not monotonically increasing (or the downward variation would be zero). Using `TestExploreTDTV`, I can see that s1 for example has a total of 257 runs - there is a slight wobble in the data.

For example, again for s1, here are ten random negative deltas between subsequent values (i.e. opposing the overall upwards trend):

```
cat s1.txt | awk 'NR==1{prev=$1; print $1, 0; next} {print $1, $1-prev; prev=$1}' | awk '{print $2}' | grep '^-' | shuf | head -n 10
-0.000790305
-0.00075579
-0.000761793
-0.000858927
-0.000887576
-0.000831506
-0.000736964
-0.00077298
-0.000748832
-0.000837645
```

The reason is likely obvious from the replicas graph:

![image copy 1.png](https://files.reviewableusercontent.io/images/reviewable/5076964/AuS7falaY_JeWdI4nIXaErxF/image.png)

s1-s4 are shuffling replicas between each other.

Note that this is much improved under `mma-only`, presumably because we're not doing these transfers between s1-s4.

I'm adding `mma-and-count` to this - it thrashes even more:

![image copy 2.png](https://files.reviewableusercontent.io/images/reviewable/5076964/AdplsL2oR_Z3v5AzxEbmoXCA/image.png)


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (approved):
Very cool :fire: I left a few questions - none are blocking. <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-:-O_XTGVyB5CisSiFcbuZ:b-xn2zdw)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola)

___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 35 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-O_XU9gTDUkFPWRiELe9:-O_XU9gTDUkFPWRiELeA:b4fa3b2) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f53afc6283553b293db1844a3857b0df2647f31/pkg/kv/kvserver/asim/history/thrashing.go#L35)):*
>  `tdtv/(normTV*T)` or `tdtv/(normTV*runs) 

Is it correct that we don’t do this yet? 
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 22 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-O_XTXMz8DfglzMcWFHy:-O_XTXMz8DfglzMcWFHz:b-12ub67) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f53afc6283553b293db1844a3857b0df2647f31/pkg/kv/kvserver/asim/history/thrashing.go#L22)):*
> ```Go
> 	// The number of runs equals one plus the number of sign changes in the first
> 	// differences.
> 	runs int
> ```

What’s the intuition behind `runs` having a plus one here? I found it a bit confusing that an empty list gives `runs = 1`. Is this for potential `tdtv/(normTV*runs)`? 
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 67 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-O_XXQis4oko9RIw0tn8:-O_XXQis4oko9RIw0tn9:b-h5g0fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f53afc6283553b293db1844a3857b0df2647f31/pkg/kv/kvserver/asim/history/thrashing.go#L67)):*
> ```Go
> 		}
> 	}
> 	_, _, normTV := extrema(values)
> ```

nit: when I was first reading this, I found it a bit confusing that we assign `normTV` here only to have it replaced in `normalize()` later. We only look up thrashing from `history.ThrashingForStat`, which will always call normalize() after having the full set of stores - is that right? But your comments above `normTV` was helpful and explain this, so it’s probably fine.
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 136 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-O_XUEEi4s0XYZ9uvFOD:-O_XUEEi4s0XYZ9uvFOE:b5if6iv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f53afc6283553b293db1844a3857b0df2647f31/pkg/kv/kvserver/asim/history/thrashing.go#L136)):*
> ```Go
> 		// log(1+lambda*r)/log(1+lambda).
> 		const p = 0.8
> 		return math.Pow(r, p)
> ```

iiuc, the formula ends up being `smaller + smaller^0.8 + larger^0.2` and smaller <= larger. What’s the intuition for favoring smaller variations when displaying the thrashing amount? I suppose that makes sense, though I’m not sure I would be able to come up with it myself.


<!-- Sent from Reviewable.io -->

wenyihu6 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152506)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @sumeerbhola and @tbg)

___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 28 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-O_X_Trj4SQfvamqooMM:-O_X_Trj4SQfvamqooMN:b3w5yid) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f53afc6283553b293db1844a3857b0df2647f31/pkg/kv/kvserver/asim/history/thrashing.go#L28)):*
> ```Go
> 	// counting mainly the movement against the trend. see tdtv for details.
> 	tdtv       float64
> 	uptv, dntv float64 // upwards and downwards total variations (both nonnegative)
> ```

nit: maybe add a comment noting that `uptv`, `dntv`, and `runs` are only used for comments/testing - could help new readers


<!-- Sent from Reviewable.io -->

tbg committed:
history: address feedback

tbg reviewed (commented):
TFTR!

I rebased and pushed a commit with some comment improvements.

bors r+

> @tbg dismissed @sumeerbhola from 6 discussions.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-:-O_keDcO0uhfaRphILYM:bb9gzjw)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @sumeerbhola and @wenyihu6)

___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 22 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-O_XTXMz8DfglzMcWFHy:-O_kbyEt-5ZGV7q5kCs0:b-bim4cq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f53afc6283553b293db1844a3857b0df2647f31/pkg/kv/kvserver/asim/history/thrashing.go#L22)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

What’s the intuition behind `runs` having a plus one here? I found it a bit confusing that an empty list gives `runs = 1`. Is this for potential `tdtv/(normTV*runs)`?
</blockquote></details>

This could go either way for the empty list, but a monotonic list has one run, so you're really asking if an empty list is monotonic. As the code shakes out, it is - but of course that's not a very interesting property.
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 28 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-O_X_Trj4SQfvamqooMM:-O_kcOi68mJX_u9SUERn:bakyqfo) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f53afc6283553b293db1844a3857b0df2647f31/pkg/kv/kvserver/asim/history/thrashing.go#L28)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

nit: maybe add a comment noting that `uptv`, `dntv`, and `runs` are only used for comments/testing - could help new readers
</blockquote></details>

I rearranged the fields and mentioned which ones are informational.
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 35 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-O_XU9gTDUkFPWRiELe9:-O_kcT_qC2dikI_EPUEm:bfn5s3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f53afc6283553b293db1844a3857b0df2647f31/pkg/kv/kvserver/asim/history/thrashing.go#L35)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

>  `tdtv/(normTV*T)` or `tdtv/(normTV*runs) 

Is it correct that we don’t do this yet?
</blockquote></details>

Yes, we don't do this yet. We currently look at thrashing across the entire series, not normalized for a unit step of time.
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 67 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-O_XXQis4oko9RIw0tn8:-O_kcjQIBbAvrmRIggdu:bx5qjz9) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f53afc6283553b293db1844a3857b0df2647f31/pkg/kv/kvserver/asim/history/thrashing.go#L67)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

nit: when I was first reading this, I found it a bit confusing that we assign `normTV` here only to have it replaced in `normalize()` later. We only look up thrashing from `history.ThrashingForStat`, which will always call normalize() after having the full set of stores - is that right? But your comments above `normTV` was helpful and explain this, so it’s probably fine.
</blockquote></details>

Yes, that's correct. Computing an individual tdtv normalizes on that series only, but in our current use case, we normalize on the min/max of the union of the series. I added a comment to `normalize` that may help future readers.
___
*[`pkg/kv/kvserver/asim/history/thrashing.go` line 136 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/152506#-O_XUEEi4s0XYZ9uvFOD:-O_kdEqFEI2L2FP7YhUF:bmeywsm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f53afc6283553b293db1844a3857b0df2647f31/pkg/kv/kvserver/asim/history/thrashing.go#L136)):*
<details><summary><i>Previously, wenyihu6 (Wenyi Hu) wrote…</i></summary><blockquote>

iiuc, the formula ends up being `smaller + smaller^0.8 + larger^0.2` and smaller <= larger. What’s the intuition for favoring smaller variations when displaying the thrashing amount? I suppose that makes sense, though I’m not sure I would be able to come up with it myself.
</blockquote></details>

You mean `smaller + larger^0.2*smaller^0.8`, right? It's a completely arbitrary choice. p=0.5 is "sort of" in the middle, but I moved it a bit higher to be a bit more aggressive at trend discounting. Again, there is no particular reason for it.


<!-- Sent from Reviewable.io -->
